driver-core/src/main/com/mongodb/internal/connection/DefaultServerMonitor.java

```diff
@@ -96,108 +113,137 @@ public void connect() {
+                    if (!connection.hasMoreToCome()) {
+                        BsonDocument ismaster = new BsonDocument("ismaster", new BsonInt32(1));
+                        if (currentServerDescription.getTopologyVersion() != null && connection.supportsAdditionalTimeout()) {
+                            ismaster.append("topologyVersion", currentServerDescription.getTopologyVersion().asDocument());
+                            ismaster.append("maxAwaitTimeMS", new BsonInt64(serverSettings.getHeartbeatFrequency(MILLISECONDS)));
+                        }
+
+                        connection.send(createCommandMessage(ismaster, connection), new BsonDocumentCodec(), sessionContext);
+                    }
+
+                    BsonDocument isMasterResult;
+                    if (currentServerDescription.getTopologyVersion() != null && connection.supportsAdditionalTimeout()) {
+                        isMasterResult = connection.receive(new BsonDocumentCodec(), sessionContext,
+                                Math.toIntExact(serverSettings.getHeartbeatFrequency(MILLISECONDS)));
+                    } else {
+                        isMasterResult = connection.receive(new BsonDocumentCodec(), sessionContext);
+                    }
+
+                    long elapsedTimeNanos = System.nanoTime() - start;
+                    serverMonitorListener.serverHeartbeatSucceeded(
+                            new ServerHeartbeatSucceededEvent(connection.getDescription().getConnectionId(), isMasterResult,
+                                    elapsedTimeNanos, currentServerDescription.getTopologyVersion() != null));
+
+                    return createServerDescription(serverId.getAddress(), isMasterResult, averageRoundTripTime.getAverage());
+                } catch (RuntimeException e) {
+                    serverMonitorListener.serverHeartbeatFailed(
+                            new ServerHeartbeatFailedEvent(connection.getDescription().getConnectionId(), System.nanoTime() - start,
+                                    currentServerDescription.getTopologyVersion() != null, e));
+                    throw e;
+                }
+            } catch (Throwable t) {
+                if (connection != null) {
+                    connection.close();
+                    connection = null;
+                }
+                return getConnectingServerDescription(t);
             }
         }
 
+        private CommandMessage createCommandMessage(final BsonDocument ismaster, final InternalConnection connection) {
+            return new CommandMessage(new MongoNamespace("admin", COMMAND_COLLECTION_NAME), ismaster,
+                    new NoOpFieldNameValidator(), primary(),
+                    MessageSettings.builder()
+                            .maxWireVersion(connection.getDescription().getMaxWireVersion())
+                            .build(),
+                    true);
```

Comment:
`true` indicates `exhaustAllowed`.  According to OP_MSG protocol, it's fine to always set this to `true` even for old servers.  It will just be ignored if streaming isn't supported.

Reply:
I tried it, and you're right, the server is not happy with the combination:

    {
      "ok": 0.0, 
     "errmsg": "An isMaster request with exhaust must specify 'maxAwaitTimeMS'", 
     "code": 51756, 
     "codeName": "Location51756"
    }

Do you think the server is acting according to specification here? If so, should I fix it or just assume that topologyVersion will always be available?

Reply:
> Do you think the server is acting according to specification here?

I can see both sides to that debate. I suggest not sending exhaustAllowed if it's not a streaming isMaster command to avoid any unexpected behavior.

Would you like to open a server ticket or shall I? It might be useful to refer back to this in the future.

Reply:
Done.  Stopped sending exhaustAllowed unless we are streaming.

I haven't opened a server ticket.  The current behavior seems acceptable.